src/custom/components/swap/EthWethWrap/index.tsx

```diff
@@ -113,160 +67,600 @@ const ButtonWrapper = styled.div`
+    wrapHash: undefined,
+  })
+
+  const isExpertMode = useIsExpertMode()
+
+  // maintain own local state of approve/wrap states
+  const [approveSubmitted, setApproveSubmitted] = useState(false)
+  const [wrapSubmitted, setWrapSubmitted] = useState(false)
+  const isApprovePending = useIsTransactionPending(pendingHashMap.approveHash)
+  const isWrapPending = useIsTransactionPending(pendingHashMap.wrapHash)
   const [nativeBalance, wrappedBalance] = useCurrencyBalances(account, [native, wrapped])
 
-  // does the user have a lower than set threshold balance? show error
-  const { isLowBalance, txsRemaining } = useMemo(
-    () => ({
-      isLowBalance: _isLowBalanceCheck({
-        threshold: multiTxCost,
-        nativeInput,
-        balance: nativeBalance,
-        txCost: singleTxCost,
+  // wrap/unwrap/native/wrapped related constants
+  const { isWrap, isUnwrap, isNative, wrappedSymbol, nativeSymbol } = useMemo(() => {
+    const isWrap = !isNativeIn && isWrappedOut
+    return {
+      isWrap,
+      isUnwrap: !isNativeOut && isWrappedIn,
+      isNative: isNativeIn || isWrap,
+      wrappedSymbol: wrapped.symbol || 'wrapped native token',
+      nativeSymbol: native.symbol || 'native token',
+    }
+  }, [isNativeIn, isNativeOut, isWrappedIn, isWrappedOut, native.symbol, wrapped.symbol])
+
+  // user safety checks to make sure any on-chain native currency operations are economically safe
+  // shows user warning with remaining available TXs if a certain threshold is reached
+  const { balanceChecks } = useRemainingTxsAndCosts({
+    native,
+    isNativeIn,
+    nativeBalance,
+    nativeInput,
+  })
+
+  // get derived EthFlow state
+  const state = useMemo(
+    () =>
+      _getDerivedEthFlowState({
+        approveError,
+        wrapError,
+        approveSubmitted,
+        wrapSubmitted,
+        isApprovePending,
+        isWrapPending,
+        needsApproval,
+        needsWrap,
+        isExpertMode,
       }),
-      txsRemaining: _getAvailableTransactions({ nativeBalance, nativeInput, singleTxCost }),
-    }),
-    [multiTxCost, nativeBalance, singleTxCost, nativeInput]
+    [
+      isExpertMode,
+      approveError,
+      approveSubmitted,
+      isApprovePending,
+      isWrapPending,
+      needsApproval,
+      needsWrap,
+      wrapError,
+      wrapSubmitted,
+    ]
   )
 
-  const wrappedSymbol = wrapped.symbol || 'wrapped native token'
-  const nativeSymbol = native.symbol || 'native token'
-
-  // Listen for changes in isWrapPending
+  // loading if either approving or wrap pending
   // and set loading local state accordingly..
   useEffect(() => {
-    setLoading(isWrapPending)
-  }, [isWrapPending])
+    setLoading(isApprovePending || isWrapPending)
+  }, [isApprovePending, isWrapPending])
+
+  const handleError = useCallback(
+    (error: any, type: 'WRAP' | 'APPROVE') => {
+      console.error(error)
+
+      if (type === 'WRAP') {
+        setWrapError(error)
+      }
+      if (type === 'APPROVE') {
```

Comment:
Nitpick:

```suggestion
      } else if (type === 'APPROVE') {
```

Reply:
Or even, since it's a required parameter and there are only 2 choices:

```suggestion
      } else {
```

Reply:
Or even shorter which for sure @anxolin will dislike:

```ts
type === 'WRAP' ? setWrapError(error) : setApproveError(error)
```

Reply:
haha yep was avoiding that (although i prefer it)